Readme.md
# Calendar-tools

  * Recurring events generation
  * Convertion functions for RFC2445 specification.
  * Designed to works on Node.JS and browsers.

## Recurring events generation

### Event structure

``` javascript
  var newEvent = {
      title: 'My next birthday'
    , start: new Date(2012, 6, 18, 16)
    , end: new Date(2012, 6, 18, 23, 30)
    , allDay: false
    , frequency: 'year'
    , recurrence: {
          'end-by': {
              after: 4
            , type: 'never'
          }
        , every: 1
        , exceptions: ['07/18/2012']
      }
   }
```

### Server-Side support
-----------------------

``` javascript
// add seed module
var Seed = require('./lib/seed');

var today = new Date();

// defines an event object
var myBirthDays = {
    title: 'Event Instances'
  , start: new Date(1977, 6, 18, 15, 30, 48)
  , end: new Date(1977, 6, 22, 15, 30, 48)
  , allDay: false
  , frequency: 'year'
  , recurrence: {
        'end-by': {
            type: 'never'
          , on: today
        }
      , every: 1
      , exceptions: []
    }
}

// creates a new seed Object passing event object and options
var Seed = new Seed(myBirthDays, {
    start: new Date(2000, 0, 1)
  , end: today
});

// generates ans retrieves all instances by period
var instances = Seed.getInstances();

for (var i = 0; i < instances.length; i++) {
  var Instance = instances[i];
  console.log(Instance.start + ': ' + Instance.getNumber() + ' years');
};
```

### building client-side library

you can use 'make build' rule to building client-side library. The generated
file will be saved in dist/calendar-tools.js

## License 

(The MIT License)

Copyright (c) 2011 Damian Suarez &lt;damian@learnboost.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.3.9 / 2014-06-05
==================

 * update root repo path

0.3.8 / 2012-01-19 
==================

  * fix getDiff() time computing bug

0.3.7 / 2012-01-19 
==================

  * fix bug for monthy(day of week) - allday recurring event

0.3.6 / 2012-01-19 
==================

  * allday correction in getElapsedInstances() method

0.3.5 / 2012-01-18 
==================

  * allday correction in getElapsedInstances() method
  * adding gd:originalEvent support to atom lib

0.3.4 / 2012-01-04 
==================

  * restoring getRecurrenceObj build proccess
  * apply new repeat-on control for jumping vector

0.3.3 / 2012-01-04 
==================

  * weekly mode: set default start day when is undefined from ical
  * change rrule detection

0.3.2 / 2011-12-21 
==================

  * incremenet 'UNTIL' value one day
  * pass day week number for monthly BYDAY recurrence mode
  * incremenet an day to TEND for all-day event

0.3.1 / 2011-12-20 
==================

  * fixing name vars errors

0.3.0 / 2011-12-20 
==================

  * fix BYMONTHDAY value

0.2.9 / 2011-12-20 
==================

  * remove time dependecy
  * add utc parameter to UTCTime string method. Cleaning
  * refact passing options for genRecurrenceString() and genEvent() methods

0.2.8 / 2011-12-20 
==================

  * fix bugs cause methods names

0.2.7 / 2011-12-20 
==================

  * add UTCTime() build date format

0.2.6 / 2011-12-19 
==================

  * add parsing ISO functions

0.2.5 / 2011-12-19 
==================

  * add parseISO8601() and toISOString() util methods

0.2.4 / 2011-12-19 
==================

  * fix variable name error.
  * remove not used functions

0.2.3 / 2011-12-08 
==================

  * add options parameter to genRecurrenceString() add genEvent()

0.2.2 / 2011-12-06 
==================

  * expose atom and rfc2445 module

0.2.1 / 2011-11-29 
==================

  * replace native Array indexOf method by util.array.indexOf function

0.2.0 / 2011-11-29 
==================

  * not use reserver 'const' word

0.1.8 / 2011-11-21 
==================

  * Fix. walking on the week.
  * add new seed weekly test files
  * changing rule for build dist

0.1.7 / 2011-11-18 
==================

  * add client-side builded file
  * add test rule to build client-side library using browserbuild
  * removing form lib/ working both-sides

0.1.6 / 2011-11-11 
==================

  * finally renames Carilu by caTools
  * updates tests

0.1.5 / 2011-11-10
=================
=
  * fix version number
  * Expose 'util'.

0.1.4 / 2011-11-07 
==================

  * Fix package.json 'main' field.
  * Renamed index to lib/calendar-tools
  * Added expresso as a developer dependency.

0.1.3 / 2011-11-05 
==================

  * add default recurrence values
  * minor improvement in applyTXCorrection()
  * apply tz correction weekly mode
  * add dylight correction in getElapsedInstance()
  * git add gitDiff(), getTZDiff() and applyTZCorrection()
  * Dylight correction when a new instance is cloned

0.1.2 / 2011-11-03 
==================

  * replace diff in days computing by _date.getDiffInDays()
  * testing elpased monthly day of week
  * fix recurrence montly - day of week.
  * add getDiff() method.
  * add defined recurrence condition when an instance is cloned
  * normalize: add recurrence default property

0.1.1 / 2011-11-02 
==================

  * monthly - day of week. date control when date number not exists in the next month
  * getDaysInMonth accept Date parameter
  * Daylight saving time nightmare correction

0.1.1 / 2011-10-31 
==================

  * store previous instance

0.1.0 / 2011-10-31 
==================

  * add client-side fullcalendar example
  * weekly - avoid falling into the void
  * getInstances() accepts start, end period dates
  * remove collection
  * doc
  * add example, API doc, etc

0.0.9 / 2011-10-31 
==================

  * normalize() - remove all-day correction
  * weekly - avoid falling into the void
  * getInstances() accepts start, end period dates

0.0.8 / 2011-10-28 
==================

  * add getSeed() method
  * addNoRec condition

0.0.7 / 2011-10-28 
==================

  * add collection property
  * integrate Seed class with Instance
  * add instance.js class

0.0.6 / 2011-10-28 
==================

  * timezone: add systemTZCorrection and offsetCorrection

0.0.5 / 2011-10-28 
==================

  * no-recurrence events setting by options param
  * process no-recurrence events
  * Readming

0.0.4 / 2011-10-27 
==================

  * add isLeapYear() and getDateInMonth() date functions
  * remove lib/index.js
  * rename test file
  * add getFormattedTime() date function()
  * fixed worng ISO8601 call
  * removign old code. Add getCountdown()
  * add/testing getElapsedInstances()
  * clearTime() clone supported

0.0.3 / 2011-10-21 
==================

  * calen is officially killed
  * atom: doc
  * RFC2445 refact: add atom lib
  * RFC2445 refact: rename getRecurrenceXML by genRecurrenceString
  * RFC2445 refact: getAllDay()
  * RFC2445 refact: remove getAllday function
  * RFC2445 refact: getFrequency() tests passed
  * RFC2445 refat: add getFrequency() function
  * RFC2445 refat: remove functions from calen to rfc2445
  * refact: add rfc2445.js new lib file.
  * refact: moving usefull parsing date functions from calen to util
  * refact: moving usefull string function from calen to util
  * util: add daysMap to date constants
  * util: exports date constants
  * util: rename getEventDuration() by getDuration()

0.0.2 / 2011-10-21 
==================

  * atom: genEvent()
  * atom: rename to genEvent() method
  * RFC2445 refact 
  * recurrence: normalized lib
  * recurrence: nromalized all tests
  * util: rename getEventDuration() by getDuration()
  * recurrence: remove util methods
  * util: renames names method

0.0.1 / 2010-01-03
==================

  * Initial release


lib/atom.js
/*!
 * atom
 * Copyright(c) 2011 Damian Suarez <rdsuarez@gmail.com>
 * MIT Licensed
 */

/**
 * Module dependecies
 */

  var Util = require('./util')
    , RFC2445 = require('./rfc2445');

/**
 * Alias
 */

  var _str = Util.string;

/**
 * genEvent() function
 * Getter that returns an event in XML format.
 * Usefull for the Google Calendar API.
 *
 * @param {Object} ev: Object Event
 *
 * return {String} event atom string
 */

module.exports.genEvent = function (ev, opts) {

  // default options
  opts = opts || {};
  var options = {
          start: ev.start
        , end: ev.end
        , timezone: 'GMT'
        , originalEvent: null
      };

  for (var k in options)
    options[k] = opts[k] || options[k];

  // timezone location
  tz = options.timezone;

  var orgEv = options,originalEvent;

  return [
      '<entry xmlns="http://www.w3.org/2005/Atom" '
    ,        'xmlns:gd="http://schemas.google.com/g/2005" '
    ,        'xmlns:gCal="http://schemas.google.com/gCal/2005" '
    ,        'gd:fields="gd:recurrence,gd:when">'
    ,   "<title>" + _str.content(ev.title) + "</title>"

    ,   ev.frequency
          ? "<gd:recurrence>"
              + _str.content(RFC2445.genRecurrenceString(ev, opts)) 
              + "</gd:recurrence>"
          : "<gd:when startTime=\""
              + _str.attr(_str.date(options.start, ev.allDay)) + "\" "
              + "endTime=\""
              + _str.attr(_str.date(options.end, ev.allDay, true)) + "\" />"

    ,   orgEv
          ? "<gd:originalEvent href=\"" + orgEv.href + "\" id=\"" + orgEv.id + "\">"
              + "<gd:when startTime=\"" + org.Ev.startTime + "\"/>"
              + "</gd:originalEvent>"
          : ""

    ,   "<gd:where valueString=\"" + _str.attr(tz) + "\"></gd:where>"
    ,   ev.isNew 
          ? "<published>" + _str.date(new Date) + "</published>"
          : "<updated>" + _str.date(new Date) + "</updated>"
    , "</entry>"
  ].join('');
};


lib/rfc2445.js
/*!
* calen
* Copyright(c) 2011 Damian Suarez <rdsuarez@gmail.com>
* MIT Licensed
*/

/**
 * Module dependecies
 */

var Util = require('./util');

/**
 * Alias
 */

var _str = Util.string
  , _date = Util.date;

/**
 * freqquency names map
 */

var freqNamesMap = {
        YEARLY: 'year'
      , MONTHLY: 'month'
      , WEEKLY: 'week'
      , DAILY: 'day'
    };

/**
 * days map from RFC2445 protocol to recurrence structure
 */

var daysMapRFC2445toRec = {
      SU: 'su'
    , MO: 'm'
    , TU: 't'
    , WE: 'w'
    , TH: 'th'
    , FR: 'f'
    , SA: 'sa'
  };

/**
 * days map from recurrence structure to RFC2445 protocol
 */

var daysMapRecToRFC2445 = {
      su: 'SU'
    , m: 'MO'
    , t: 'TU'
    , w: 'WE'
    , th: 'TH'
    , f: 'FR'
    , sa: 'SA'
  };

/**
 * days names
 */

var dayNames = ['su', 'm', 't', 'w', 'th', 'f', 'sa'];

/**
 * RFC2445 usefull functions
 */

var RFC2445 = module.exports = {};

/**
 * generate functions *
 */

/**
 * genRecurrenceString
 * generates a RFC2445 recurrence string passing an Event Object
 *
 * @param {Object} ev: Event Object
 *
 * return {String} RFC2445 recurrence string
 */

RFC2445.genRecurrenceString = function (ev, opts) {
  if (!ev.frequency) return '';

  opts = opts  || {};

  var options = {
          start: ev.start
        , end: ev.end
        , timezone: 'GMT'
      };

  for (var k in options)
    options[k] = opts[k] || options[k];

  var tz = options.timezone;

  // period: DTSTART, DTEND
  var rules = [
          'DTSTART' + (ev.allDay
                          ? ';' + _str.value(options.start)
                          : ';' + _str.dateTZID(options.start, tz))
        , 'DTEND' + (ev.allDay
                          ? ';' + _str.value(options.end, false, ev.allDay)
                          : ';' + _str.dateTZID(options.end, tz))
      ];

  // RRULE - frequency value
  var rrule = ''
    , steps = ['freq', 'repeat', 'end-by', 'interval']
    , endBy = ''
    , interval = '';

  // end-by type

  // COUNT
  if (ev.recurrence['end-by'].type == 'after')
    endBy = ';COUNT=' + ev.recurrence['end-by'].after;
  // ON
  if (ev.recurrence['end-by'].type == 'on')
    endBy = ';UNTIL=' + _str.value(ev.recurrence['end-by'].on, true, true);

  // INTERVAL
  if (ev.recurrence.every != 1)
    interval = ';INTERVAL=' + ev.recurrence.every;

  for (var st = 0; st < steps.length; st++) {
    if (st == 0 || st == 2) {
      switch (ev.frequency) {
        case 'year':
          rrule += st == 0
                    ? 'RRULE:FREQ=YEARLY'
                    : '';
        break;

        case 'month':
          rrule += st == 0
                    ? 'RRULE:FREQ=MONTHLY'
                    : (st == 2
                        ? (ev.recurrence['repeat-on'].mode == 'day-of-month'
                            ? ';BYMONTHDAY=' + ev.start.getDate()
                            : ';BYDAY=' + (_date.getWeekOfDay(ev.start) + 1)
                              + daysMapRecToRFC2445[dayNames[options.start.getDay()]])
                        : '');
        break;

        case 'week':
          var days = '';
          for (var l in ev.recurrence['repeat-on'])
            days += daysMapRecToRFC2445[l] + ','

          days = days.substring(0, days.length - 1);

          rrule += st == 0
                    ? 'RRULE:FREQ=WEEKLY;BYDAY=' + days
                    : '';
        break;

        case 'day':
          rrule += st == 0
                    ? 'RRULE:FREQ=DAILY'
                    : '';
        break;
      };
    }

    // add end-by
    else if (st == 1)
      rrule += endBy;

    // add end-by
    else if (st == 3)
      rrule += interval;

  };

  rules.push(rrule);

  return rules.join("\r\n");
}

/**
 * getting functions
 */

/**
 * getRecurrenceObj() method.
 * return a recurrence object of LearnBoost event passing RFC2445 string
 *
 * @param {string} str: RFC2445 recurrence string
 * return {Object} recurrence LearnBoost object
 */

RFC2445.getRecurrenceObj = function (str) {
  // debug: RFC2445 \x1B[0;33m%s\x1B[0m string to recurrence object, str
  var recRule = str.match(/RRULE(.+)/)
    , recObj = {
          every: 1
        , 'end-by': {}
        , exceptions: []
        , 'repeat-on': {}
      }
    , _freq = '';

  recRule = recRule.length ? recRule[0] : null;

  if (!recRule) return null;

  // debug: recurrence rule \x1B[0;33m%s\x1B[0m, recRule
  var recPrpsRule = recRule.split(';');

  for (var k = 0; k < recPrpsRule.length; k++) {
    var prp = recPrpsRule[k].split('=');
    // debug: recurrence property [%s]: \x1B[0;33m%s: %j\x1B[0m, k, prp[0], prp[1]

    // recurrence type
    if (prp[0] == 'RRULE:FREQ') {
      _freq = prp[1];
    }

    // INTERVAL/repeat
    if (prp[0] == 'INTERVAL') {
      recObj.every = prp[1];
    }

    // end-by properties
    // AFTER mode
    else if (prp[0] == 'COUNT') {
      recObj['end-by'].type = 'after';
      recObj['end-by'].after = prp[1];
    }

    // UNTIL/on mode
    else if (prp[0] == 'UNTIL') {
      recObj['end-by'].type = 'on';
      var data = _date.parseFromWeirdISOFormat(prp[1]);
      recObj['end-by'].on = new Date(data);
      
    }

    // BYDAY/repeat-on. Used for:
    // - weekly mode
    // - monthly mode - day on week
    else if (prp[0] == 'BYDAY') {

      if (_freq == 'WEEKLY') {
        var days = prp[1].split(',');
        for (var l = 0; l < days.length; l++) {
          var key = daysMapRFC2445toRec[days[l]];
          recObj['repeat-on'][key] = 'on';
        };
      }
      else if (_freq == 'MONTHLY') {
        // for monthly trecurrence type simply defines mode
        recObj['repeat-on'].mode = 'day-of-week';
        
      }
    }

    // BYMONTHDAY/repeat-on. Used for:
    // - monthly mode
    else if (prp[0] == 'BYMONTHDAY')
      recObj['repeat-on'].mode = 'day-of-month';

  };

  // setting default options
  // end-by type
  recObj['end-by'].type = recObj['end-by'].type || 'never';

  return recObj;
};

/**
 * getFrequency
 * retrieve frequency property from google RFC2445 recurrence string
 *
 * @param {string} str
 * @param {Object} frequency names map
 *    {
 *        YEARLY: 'year'
 *      , MONTH:  'month'
 *      , WEEKLY: 'week'
 *      , DAILY': 'day'
 *    }
 */

RFC2445.getFrequency = function (str, namesMap) {
  // debug: get frequency property from RFC2445 \x1B[0;33m%s\x1B[0m, str
  // frequency names map
  namesMap = namesMap || freqNamesMap;
 
  var frequency = str.match(/FREQ=([A-Z]+)/);

  if (frequency)
    return namesMap[frequency[1]];
  return null;
};

/**
 * retrieve all-day property from RFC2445 recurrence string
 */

RFC2445.getAllday = function (str) {
  // debug: get all-day property from RFC2445 \x1B[0;33m%s\x1B[0m, str
  // deduces 'all-day' property through the DTSTART format property
  var dtstart = str.match(/(^DTSTART;([^ ]+))(:\w+)(T)/);

  return dtstart ? false : true;
};


lib/instance.js
/*!
* Instance
* Copyright(c) 2011 Damian Suarez <rdsuarez@gmail.com>
* MIT Licensed
*/

/**
 * add util module
 */

var util = require('./util');

/**
 * alias
 */

var _date = util.date;

/**
 * exports Instance Class
 */

var modInstance = module.exports = function (Seed, options) {
  return new Instance(Seed, options);
}

/**
 * Instance class. Creates a instance of seed event
 */

function Instance (Seed, dateReference) {
  this._seed = Seed;
  this._create(dateReference);
}

// alias Instance prototype
var pptInst = Instance.prototype;

/**
 * creates an instance event
 */

pptInst._create = function (dateReference) {
  var clone = this.clone()
    , diff = _date.getDiff(this._seed.ev.end, this._seed.ev.start);

  // set cloned start date
  clone.start = new Date(
          dateReference.getFullYear()
        , dateReference.getMonth()
        , dateReference.getDate()
        , this._seed.ev.start.getHours()
        , this._seed.ev.start.getMinutes()
        , this._seed.ev.start.getSeconds()
        , this._seed.ev.start.getMilliseconds()
    );

  // set cloned end date
  clone.end = new Date(+clone.start + diff);
  _date.applyTZCorrection(clone.end, clone.start);

  // set cloned instances as Instance Object properties
  for (var k in clone)
    this[k] = clone[k];

  return clone;
}

/**
 * clone an seed event
 */

pptInst.clone = function () {
  var newInstance = JSON.parse(JSON.stringify(this._seed.ev));

  newInstance.start = new Date(newInstance.start);
  newInstance.end = new Date(newInstance.end);

  if (newInstance.stop)
    newInstance.stop = _date.parseISO8601(newInstance.stop);

  // on date generation
  if (newInstance.recurrence) {
    var endBy = newInstance.recurrence['end-by'];
    if (endBy && endBy.on)
      newInstance.recurrence['end-by'].on = new Date(endBy.on);
  }

  return newInstance;
};

/**
 * get the instance number
 */

pptInst.getNumber = function () {
  return this.n;
}


/**
 * get if is the last instance into series
 */

pptInst.isLast = function () {
  var c = this.getSeed()._getCountdown();
  return this.getNumber() === (c - 1);
}

/**
 * get the previous instance into series
 */

pptInst.getPrev = function () {
  return this._prevInst;
}

/**
 * get the seed event
 */

pptInst.getSeed = function () {
  return this._seed;
}


lib/recurrence.js
/*!
* recurrence
* Copyright(c) 2011 Damian Suarez <rdsuarez@gmail.com>
* MIT Licensed
*/

/**
 * modeule dependencies
 */
var Seed = require('./seed');

/**
 * exports recurrence Class
 */

var modRecurrence = module.exports = function (evs, options) {
  return new Recurrence(evs, options);
}

/**
 * Initialize a new 'Recurrence' object with the given 'ev'
 *
 * @param {Object|Array} evs: object event or an array events collection
 * @param {Object}  options: {
 *      period: {
 *          start: {Date} - period start date
 *        , end: {Date} - period end date
 *      }
 *    , addNoRec: {Boolean} - add no-recurrence events
 *    , systemTZCorrection: {Boolean} - apply system timezone correction
 *    , offsetCorrection: {Number} - apply offset correction (minutes)
 *  }

 * return {Array} instances series
 */

function Recurrence (evs, options) {
  // add default options
  var start = new Date()
  this.options = {
          start: start
        , end: new Date(+start + 1000*60*6024)
        , addNoRec: true
        , systemTZCorrection: false
        , offsetCorrection: 0
      };

  // mix options
  for (var k in options)
    this.options[k] = options[k];

  // store events
  evs = evs instanceof Array ? evs : [evs];
  this.events = evs;

  // set period
  this.setPeriod();

  // planting seeds
  this.seeds = [];
  this.plantingSeeds();
}

// proto alias
var _pttRec = Recurrence.prototype;

/**
 * setPeriod() method.
 */

_pttRec.setPeriod = function (start, end) {
  start = start || this.options.start || new Date();
  end = end || this.options.end;

  // set period
  if (!end) {
    end = new Date(+start);
    end.setDate(start.getDate() + 1);
  }

  var p = this.period = {
          start: typeof start == 'string' ? new Date(start) : start
        , end: typeof end == 'string' ? new Date(end) : end
      };

  return p;
}

/**
 * planting seeds. That is register all events crating 'seed' objects
 */

_pttRec.plantingSeeds = function () {
  for (var s = 0; s < this.events.length; s++)
    this.seeds.push(new Seed(this.events[s], this, this.options));
}

/**
 * return seeed events
 */

_pttRec.getSeeds = function () {
  return this.seeds;
}

/**
 * getInstances()
 *
 * return {Array} all instances of all seed events
 */

_pttRec.getInstances = function (start, end) {
  var allInstances = [];

  for (var i = 0; i < this.seeds.length; i++) {
    var seed = this.seeds[i];
    allInstances = allInstances.concat(seed.getInstances(start, end));
  };

  return allInstances;
}


lib/util.js
/*!
* util usefull functions
* Copyright(c) 2011 Damian Suarez <rdsuarez@gmail.com>
* MIT Licensed
*/

/**
 * time constants
 */

var _minute = 1000 * 60
  , _hour = _minute*60
  , _day = _hour*24
  , _week = _day * 7;

/**
 * days map
 * used in weekly mode process.
 */

var daysMap = { su: 0, m: 1, t: 2, w: 3, th: 4, f: 5, sa: 6 };

/**
 * date usefull functions *
 */

var _date = module.exports.date = {};

/**
 * exports constants
 */

_date['const'] = {
    minute: _minute
  , hour: _hour
  , day: _day
  , week: _week
  , daysMap: daysMap
}

/**
 * clearTime()
 *
 * @param {Date} date: date to clear time
 * @para, {Boolean} clone: if is true clones new date
 *
 * return {Date} date
 */

var clearTime = _date.clearTime = function (date, clone) {
  var _date = date;

  if (clone)
    _date = new Date(date);

  else if (typeof date == 'number')
    _date = new Date(Number(date));

  _date.setMilliseconds (0);
  _date.setSeconds (0);
  _date.setMinutes (0);
  _date.setHours (0);

  return _date;
}

/**
 * Returns if a certain year is a leap year
 */

_date.isLeapYear = function (year){
  return (((year % 4 === 0) && (year % 100 !== 0)) || (year % 400 === 0));
};

/**
 * Returns days in a month by year
 */

_date.getDaysInMonth = function (){
  var args = arguments
    , year = args[0] instanceof Date ? args[0].getFullYear() : args[0]
    , month = args[0] instanceof Date ? args[0].getMonth() : args[1];

  return [31, (_date.isLeapYear(year) ? 29 : 28), 31, 30, 31, 
          30, 31, 31, 30, 31, 30, 31][month];
};

/**
 * Apply timezone diff correction
 */

_date.applyTZCorrection = function (date, dateToCompare, sign) {
  sign = sign || 1;
  var tzdiff = _date.getTZDiff(date, dateToCompare);

  if (tzdiff)
    date.setMinutes(date.getMinutes() + tzdiff * sign);

  return date;
}

/**
 * getTZDiff()
 * returns the timezone difference between given dates
 */

_date.getTZDiff = function (end, start) {
  return end.getTimezoneOffset() - start.getTimezoneOffset()
}

/**
 * diff in milliseconds
 * support Dylight saving time nightmare correction.
 */

_date.getDiff = function (end, start) {
  var dim = _date.getTZDiff(end, start) * _minute
    , diff = +end - +start - dim;
  return diff;
}

/**
 * diff in days
 * support Dylight saving time nightmare correction.
 */

_date.getDiffInDays = function (end, start, mtd) {
  var diff = _date.getDiff(end, start);
  mtd = mtd || 'ceil';
  return Math[mtd](diff / _day);
}


/**
 * getDiffInMonths()
 */

_date.getDiffInMonths = function (end, start, noRound) {
  var tempStart = new Date(start)
    , tempEnd = new Date(end);

  tempStart.setFullYear(2000);
  tempEnd.setFullYear(2000);

  tempStart.setMonth(0);
  tempEnd.setMonth(0);

  var less = tempEnd < tempStart && noRound ? 1 : 0;
  return (end.getFullYear() - start.getFullYear()) * 12
       + (end.getMonth() - start.getMonth() - less);
}

/**
 * getDiffInMonths()
 */

_date.getDiffInYears = function (end, start, noRound) {
  return _date.getDiffInMonths(end, start, noRound) / 12 | 0;
}

/**
 * getFormattedDate()
 *
 * @paramn {Date} date
 * @param {Boolean} cp: complete format date with zeros
 *
 * return {String} date: eg 07/18/1977
 */

var getFormattedDate = _date.getFormattedDate = function (date, cp) {
  var d = date.getDate()
    , m = date.getMonth() + 1
  var arrDate = [
          cp && m < 10 ? '0' + m : m
        , cp && d < 10 ? '0' + d : d
        , date.getFullYear()
      ]

  return arrDate.join('/');
}

/**
 * getFormattedTime()
 * return hh:mm tt string
 */

var getFormattedTime = _date.getFormattedTime = function (date) {
  var h = date.getHours()
    , m = date.getMinutes()
    , s = date.getSeconds()
    , ampm =  h < 12 ? 'AM' : 'PM';

  h = !h ? 12 : (h > 12 ? (h - 12) : h);

  return (h < 10 ? '0' + h : h) + ':' + (m < 10 ? '0' + m : m) + ' ' + ampm;
}

/**
 * Returns week number for the date into current view
 * @param {Date Object} date
 * @return {Number}
 */

var getWeekOfDay = _date.getWeekOfDay = function (date) {
  return Math.floor((date.getDate() - 1 ) / 7);
}

/**
 * setByWeekAndDay()
 *
 * set date passing week and day number.
 * @param {Date} date: date to modify
 * @param {String|Number} week: week number (row)
 * @param {String|Number} day: day number (col)
 *
 * return {Date}
 *
 * eg: Oct 2011
 *
 *     S   M   T   W  Th   F  Sa
 *     0   1   2   3   4   5   6 
 *  ------------------------------
 *  | 25  26  27  28  29  30   1 |
 *  |  2   3   4   5   6   7   8 |
 *  |  9  10  11  12  13  14  15 |
 *  | 16  17  18  19  20  21  22 |
 *  | 23  24  25  26  27  28  29 |
 *  | 30  31   1   2   3   4   5 |
 *  ------------------------------
 *
 * week number is always into date month
 *
 * var d = new Date (2011, 9, 1);
 * setByWeekAndDay (d, 0, 6) -> Sat, 01 Oct 2011
 * setByWeekAndDay (d, 0, 1) -> Mon, 04 Oct 2011
 * setByWeekAndDay (d, 2, 4) -> Thu, 20 Oct 2011
 *
 */

var setByWeekAndDay = _date.setByWeekAndDay = function (date, week, day) {
  var d = new Date(date);
  d.setDate(1);

  var diff = d.getDay() - day
    , n = (week + 1) * 7 - diff + 1;

  n = diff <= 0 ? n - 7 : n;

  d.setDate(n);
  if (_date.getDiffInMonths(d, date))
    n -= 7;

  date.setDate(n);

  return date;
}


/**
 * fixDate
 * copied from fullCalendar
 */

function fixDate(d, check) {
  // force d to be on check's YMD, for daylight savings purposes
  if (+d) { // prevent infinite looping on invalid dates
    while (d.getDate() != check.getDate()) {
      d.setTime(+d + (d < check ? 1 : -1) * HOUR_MS);
    }
  }
}

/**
 * parseISO8601 - copied form fullCalendar
 */

var parseISO8601 = _date.parseISO8601 = function (s, ignoreTimezone) {
 if (typeof s != 'string')
    return s;

  // ignoreTimezone defaults to false
  // derived from http://delete.me.uk/2005/03/iso8601.html
  // TODO: for a know glitch/feature, read tests/issue_206_parseDate_dst.html
  var m = s.match(/^([0-9]{4})(-([0-9]{2})(-([0-9]{2})([T ]([0-9]{2}):([0-9]{2})(:([0-9]{2})(\.([0-9]+))?)?(Z|(([-+])([0-9]{2})(:?([0-9]{2}))?))?)?)?)?$/);
  if (!m)
    return null;

  var date = new Date(m[1], 0, 1);

  if (ignoreTimezone || !m[14]) {
    var check = new Date(m[1], 0, 1, 9, 0);
    if (m[3]) {
      date.setMonth(m[3] - 1);
      check.setMonth(m[3] - 1);
    }
    if (m[5]) {
      date.setDate(m[5]);
      check.setDate(m[5]);
    }
    fixDate(date, check);
    if (m[7]) {
      date.setHours(m[7]);
    }
    if (m[8]) {
      date.setMinutes(m[8]);
    }
    if (m[10]) {
      date.setSeconds(m[10]);
    }
    if (m[12]) {
      date.setMilliseconds(Number("0." + m[12]) * 1000);
    }
    fixDate(date, check);
  }
  else {
    date.setUTCFullYear(
        m[1]
      , m[3] ? m[3] - 1 : 0
      , m[5] || 1
    );
    date.setUTCHours(
        m[7] || 0
      , m[8] || 0
      , m[10] || 0
      , m[12] ? Number("0." + m[12]) * 1000 : 0
    );

    var offset = Number(m[16]) * 60 + (m[18] ? Number(m[18]) : 0);
    offset *= m[15] == '-' ? 1 : -1;
    date = new Date(+date + (offset * 60 * 1000));
  }

  return date;
}

/**
 * parse from weird ISO format
 */

_date.parseFromWeirdISOFormat = function (d, tz) {
  var parts = d.match(/(\d\d\d\d)(\d\d)(\d\d)T(\d\d)(\d\d)(\d\d)/);

  if (!parts) {
    var simple = d.match(/(\d\d\d\d)(\d\d)(\d\d)/);
    if (!simple) return;

    return new Date(simple[1], simple[2] -1, simple[3]);
  }

  var date = new Date();

  date.setMilliseconds(0);
  date.setSeconds(parts[6]);
  date.setMinutes(parts[5]);
  date.setHours(parts[4]);

  date.setDate(parts[3]);
  date.setMonth(parts[2] - 1);
  date.setFullYear(parts[1]);

  return date;
};

/**
 * parse date from TZID format
 */

_date.parseFromTZID = function (v) {
  var parts = v.match(/TZID=([^:]+):(\d+)T(\d+)/);
  if (!parts) return;

  var date = _date.parseFromWeirdISOFormat(parts[2] + 'T' + parts[3], parts[1]);

  // debug: TZID parsed with date %s, date
  return date;
};

/**
 * String usefull functions
 */

var _string = module.exports.string = {};

/**
 * Escape a string with html entities
 *
 * @param {String} string to escape
 * @api public
 */


_string.escape = function (str){
  return str
    .replace(/&/g, '&amp;')
    .replace(/>/g, '&gt;')
    .replace(/</g, '&lt;')
    .replace(/"/g, '&quot;');
};

_string.attr = function (str) {
  return str ? str.replace(/"/g, '\\"') : '';
};

_string.content = function (str) {
  return str ? _string.escape(str) : '';
};

/**
 * date method
 * @param {Date} v: date object
 * @param {Boolean} a: all-day
 * @param {Boolean} oc: date offset correction
 *
 * oc(offset correction) is used when the object event is patched to google
 * event. We need apply one day to make compatibily between fullCalendar and
 * google.
 *
 */

_string.date = function (v, a, oc) {
  // clone date
  var _v = new Date(v);
  if (a && oc)
    _v.setDate(_v.getDate() + 1);

  var date = JSON.stringify(_v).replace(/"/g, '');
  return a ? date.substring(0, 10) : date;
};

_string.dateTZID = function (v, tz) {
  var str = (JSON.stringify(v).replace(/[-, :,\.]/g, '')).substring(1, 16);
  return 'TZID=' + (tz ? (tz + ':' + str) : str);
};

_string.UTCTime = function (v, utc) {
  var str = (JSON.stringify(v).replace(/[-, :,\.]/g, '')).substring(1, 16);
  str += utc ? 'Z' : '';

  return str;
};


/**
 * value() method
 * return value date adding (or not) txt label
 *
 * @param {Date} _d
 * @param {Boolean} noAddLabel
 * @param {Boolean} allDay correction
 */

_string.value = function (_d, noAddLabel, allDay) {
  var d = new Date(_d);
  if(allDay) d.setDate(d.getDate() + 1);

  var strDate = JSON.stringify(d).split('T')[0].replace(/[", -]/g, '');
  return !noAddLabel ? 'VALUE=DATE:' + strDate : strDate;
};


/**
 * Array usefull functions
 */

var _array = module.exports.array = {};

/**
 * indexOf(). supports for IE
 *
 * @param {Array} array
 * @param item
 */

_array.indexOf = function (array, item) {
  if (!Array.prototype.indexOf) {
    /*
     * copied from Mozilla Develop Center
     * https://developer.mozilla.org/en/JavaScript/Reference/Global_Objects/Array/indexO://developer.mozilla.org/en/JavaScript/Reference/Global_Objects/Array/indexOf
     */

    if (array === void 0 || array === null)
      throw new TypeError();

    var t = Object(array)
      , len = t.length >>> 0;

    if (len === 0)
      return -1;

    var n = 0;
    if (arguments.length > 1) {
      n = Number(arguments[2]);
      if (n !== n)
        n = 0;
      else if (n !== 0 && n !== Infinity && n !== -Infinity)
        n = (n > 0 || -1) * Math.floor(Math.abs(n));
    }
    if (n >= len)
      return -1;

    var k = n >= 0 ? n : Math.max(len - Math.abs(n), 0);
    for (; k < len; k++)
      if (k in t && t[k] === item)
        return k;

    return -1;
  }
  else
    return array.indexOf(item);
}
